crates/arroyo-connectors/src/iggy/source/mod.rs

```diff
@@ -0,0 +1,314 @@
+                                return Ok(SourceFinishType::Graceful);
+                            }
+                            StopMode::Immediate => {
+                                return Ok(SourceFinishType::Immediate);
+                            }
+                        }
+                    }
+                    ControlMessage::LoadCompacted { compacted } => {
+                        ctx.load_compacted(compacted).await;
+                    }
+                    ControlMessage::NoOp => {}
+                    ControlMessage::Commit { .. } => {
+                        unreachable!("Sources shouldn't receive commit messages");
+                    }
+                }
+            }
+
+            let stream_id = Identifier::from_str(&self.stream).map_err(|e| {
+                UserError::new("Invalid stream identifier", format!("Error: {:?}", e))
+            })?;
+
+            let topic_id = Identifier::from_str(&self.topic).map_err(|e| {
+                UserError::new("Invalid topic identifier", format!("Error: {:?}", e))
+            })?;
+
+            let polling_strategy = PollingStrategy::offset(offset);
+
+            let messages_per_batch = 100;
+
+            let polled_messages = match self
+                .client
+                .as_ref()
+                .unwrap()
+                .poll_messages(
+                    &stream_id,
+                    &topic_id,
+                    Some(self.partition_id),
+                    &consumer,
+                    &polling_strategy,
+                    messages_per_batch,
+                    self.auto_commit,
+                )
+                .await
+            {
+                Ok(messages) => messages,
+                Err(e) => {
+                    error!("Error polling messages: {:?}", e);
+
+                    sleep(Duration::from_millis(1000)).await;
+                    continue;
+                }
+            };
+
+            if polled_messages.messages.is_empty() {
+                sleep(Duration::from_millis(100)).await;
+                continue;
+            }
+
+            for message in &polled_messages.messages {
+                offset = message.offset + 1;
+
+                self.process_message(message, collector).await?;
+
+                if rate_limiter.check().is_err() {
+                    sleep(Duration::from_millis(10)).await;
+                }
+            }
+
+            if collector.should_flush() {
+                collector.flush_buffer().await?;
+            }
+        }
+    }
+
+    async fn process_message(
+        &self,
+        message: &PolledMessage,
+        collector: &mut SourceCollector,
+    ) -> Result<(), UserError> {
+        let timestamp =
+            { SystemTime::UNIX_EPOCH + Duration::from_millis(message.timestamp / 1000) };
+
+        let mut additional_fields = HashMap::new();
+        for field in &self.metadata_fields {
+            match field.key.as_str() {
+                "offset" => {
```

Comment:
This can be made a bit more succinct by pulling the insert call out of each match statement (since it's always the same)